File: native/src/segments.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const label = @import("label.zig");

pub const FLAG_SHADE: u32 = 1 << 16;
pub const FLAG_RIGHT_EDGE: u32 = 1 << 17;
pub const FLAG_RISING_EDGE: u32 = 1 << 18;
pub const FLAG_FALLING_EDGE: u32 = 1 << 19;
pub const FLAG_MUTE: u32 = 1 << 20;
pub const FLAG_RISING_EDGE_LEFT: u32 = 1 << 21;
pub const FLAG_FALLING_EDGE_LEFT: u32 = 1 << 22;

// Segment is now lean: timing (t_start, t_end) + row_flags (row index + edge/shade
// bits). 3 × u32 = 12 bytes. The actual bit values live in shared x0/x1 pools,
// indexed via RowInfo (sample index = instance_index - RowInfo.segment_start).
pub const PackedSegment = extern struct {
    t_start: u32,
    t_end: u32,
    row_flags: u32,
};

pub const PACKED_SEGMENT_BYTES: usize = @sizeOf(PackedSegment);

// Per-row metadata. bytes_per_sample is ceil(bit_width / 8) (= tide's
// Type.bytes()): each sample occupies that many consecutive bytes in the pool —
// tide's native per-sample byte run, little-endian, memcpy'd straight in (no word
// repacking). x0_offset/x1_offset are BYTE offsets into the (u32-typed) pools.
// segment_start is this row's first instance index within its pipeline; sample
// index for instance ii of this row = ii - segment_start.
pub const RowInfo = extern struct {
    x0_offset: u32,
    x1_offset: u32,
    bytes_per_sample: u32,
    segment_start: u32,
    // Per-row render flags (bit 0 = dim; see ROW_FLAG_DIM in digital.wgsl /
    // digital.ts). Native always emits 0 — the renderer sets this directly in
    // the GPU rowInfo buffer (eye toggle) without a repack.
    flags: u32,
    // Per-row vertical placement, CSS px stored as f32 bits (the shader bitcasts
    // them). y_offset = the row's top in canvas space, height = its drawn height.
    // Native always emits 0 — the renderer writes the live layout into the GPU
    // rowInfo buffer (row resize) without a repack, same as `flags`.
    y_offset: u32,
    height: u32,
};

pub const ROW_FLAG_DIM: u32 = 1 << 0;

pub const ROW_INFO_BYTES: usize = @sizeOf(RowInfo);

// Number of bytes each sample of a `width`-bit signal occupies in the pools —
// tide's native stride (Type.bytes() = ceil(width/8)), memcpy'd directly.
pub fn bytesPerSample(width: u32) u32 {
    std.debug.assert(width >= 1);
    return (width + 7) / 8;
}

// ceil(width / 32) — the u32-word count used ONLY by the CPU value path
// (main.zig getValueAt / jsWordArray), which is independent of the GPU pools and
// keeps its word-array return shape. Not used by the byte-stride pools.
pub fn wordsPerSample(width: u32) u32 {
    std.debug.assert(width >= 1);
    return (width + 31) / 32;
}

const RowAccum = struct {
    bit_width: u32 = 0, // 0 = unused row
    segment_start: u32 = 0, // first instance index in this row's pipeline
    started: bool = false,
    count: u32 = 0, // number of samples pushed (lsbs/msbs hold count·bytes_per_sample)
    lsbs: std.ArrayList(u8) = .empty,
    msbs: std.ArrayList(u8) = .empty,

    fn deinit(self: *RowAccum, gpa: Allocator) void {
        self.lsbs.deinit(gpa);
        self.msbs.deinit(gpa);
    }
};

pub const MAX_ROWS: usize = 64;

// One fully packed signal, independent of its row placement — the cacheable unit
// (main.zig keys these by signal+config and reuses them across repacks, so an
// add/remove/reorder only queries tide + formats labels for the changed signal).
// `row_flags` here have the low 16 bits (row index) zeroed; the row is OR'd in at
// assembly (pushPackedSignal). `lsbs`/`msbs` hold segments.len·bytes_per_sample
// bytes (tide's raw byte planes, copied in one memcpy by setSamples).
// `label_offsets` holds segments.len+1 prefix offsets when `is_multi` and a
// segment exists (label i = label_bytes[off[i]..off[i+1]]), else stays empty.
pub const PackedSignal = struct {
    is_multi: bool = false,
    bit_width: u32 = 0,
    segments: std.ArrayList(PackedSegment) = .empty,
    lsbs: std.ArrayList(u8) = .empty,
    msbs: std.ArrayList(u8) = .empty,
    label_bytes: std.ArrayList(u8) = .empty,
    label_offsets: std.ArrayList(u32) = .empty,

    pub fn deinit(self: *PackedSignal, gpa: Allocator) void {
        self.segments.deinit(gpa);
        self.lsbs.deinit(gpa);
        self.msbs.deinit(gpa);
        self.label_bytes.deinit(gpa);
        self.label_offsets.deinit(gpa);
    }

    // Append one transition's timing+flags header (called by pack.packSignal).
    // `flags` must NOT carry row bits. Sample bytes are NOT copied here — they're
    // bulk-copied once via setSamples (tide's whole byte plane in one memcpy).
    pub fn pushSegment(self: *PackedSignal, gpa: Allocator, t_start: u32, t_end: u32, flags: u32) !void {
        try self.segments.append(gpa, .{ .t_start = t_start, .t_end = t_end, .row_flags = flags });
    }

    // Copy tide's full per-signal byte planes (x0s/x1s, len·bytes_per_sample bytes
    // each) verbatim into the value pools — the single memcpy that replaces the old
    // per-sample byte→word repack. Call once, after all pushSegment calls; the i-th
    // bytes_per_sample-byte run lines up with segment i.
    pub fn setSamples(self: *PackedSignal, gpa: Allocator, x0s: []const u8, x1s: []const u8) !void {
        try self.lsbs.appendSlice(gpa, x0s);
        try self.msbs.appendSlice(gpa, x1s);
    }

    // Append this segment's value label (multi-bit rows only). Call once per
    // pushSegment, in order. Muted segments get an empty label.
    pub fn pushLabel(
        self: *PackedSignal,
        gpa: Allocator,
        x0: []const u8,
        x1: []const u8,
        radix: label.Radix,
        enums: []const label.EnumEntry,
        muted: bool,
    ) !void {
        if (self.label_offsets.items.len == 0) try self.label_offsets.append(gpa, 0);
        if (!muted) try label.formatValue(&self.label_bytes, gpa, x0, x1, self.bit_width, radix, enums);
        try self.label_offsets.append(gpa, @intCast(self.label_bytes.items.len));
    }
};

pub const Scene = struct {
    gpa: Allocator,
    multi: std.ArrayList(PackedSegment) = .empty,
    single: std.ArrayList(PackedSegment) = .empty,
    // Per multi-bit segment value label, formatted natively (label.zig) so the
    // renderer needs no per-segment getValueAt + JS formatting. `multi_label_bytes`
    // is the concatenated ASCII; `multi_label_offsets` holds multiCount+1 prefix
    // offsets aligned with `multi` (label i = bytes[off[i]..off[i+1]]).
    multi_label_bytes: std.ArrayList(u8) = .empty,
    multi_label_offsets: std.ArrayList(u32) = .empty,
    rows: [MAX_ROWS]RowAccum = [_]RowAccum{.{}} ** MAX_ROWS,

    pub fn init(gpa: Allocator) Scene {
        return .{ .gpa = gpa };
    }

    pub fn deinit(self: *Scene) void {
        self.multi.deinit(self.gpa);
        self.single.deinit(self.gpa);
        self.multi_label_bytes.deinit(self.gpa);
        self.multi_label_offsets.deinit(self.gpa);
        for (&self.rows) |*r| r.deinit(self.gpa);
    }

    // Place an already-packed signal at `row`: append its segments to `target`
    // (OR'ing the row into each row_flags), its samples to the row's pools, and its
    // labels to the multi label buffers. This is the cache-replay path — no tide
    // query, no flag recompute, no label format. `target` must be `multi` for an
    // is_multi signal, `single` otherwise.
    pub fn pushPackedSignal(self: *Scene, target: *std.ArrayList(PackedSegment), row: u32, ps: *const PackedSignal) !void {
        std.debug.assert(row < MAX_ROWS);
        if (ps.segments.items.len == 0) return; // empty signal contributes no row data
        const bps = bytesPerSample(ps.bit_width);
        var ra = &self.rows[row];
        // Each row is filled by exactly one signal, contiguously.
        std.debug.assert(!ra.started);
        ra.bit_width = ps.bit_width;
        ra.segment_start = @intCast(target.items.len);
        ra.started = true;

        for (ps.segments.items) |s| {
            try target.append(self.gpa, .{
                .t_start = s.t_start,
                .t_end = s.t_end,
                .row_flags = (s.row_flags & ~@as(u32, 0xffff)) | (row & 0xffff),
            });
        }
        // One memcpy of the signal's whole byte run into the row's pool (replaces
        // the per-segment slice copy). One signal fills a row, so ra starts empty.
        try ra.lsbs.appendSlice(self.gpa, ps.lsbs.items);
        try ra.msbs.appendSlice(self.gpa, ps.msbs.items);
        ra.count += @intCast(ps.segments.items.len);
        std.debug.assert(ra.lsbs.items.len == ra.count * bps); // guards double-copy / stride drift

        if (ps.is_multi) {
            if (self.multi_label_offsets.items.len == 0) try self.multi_label_offsets.append(self.gpa, 0);
            for (0..ps.segments.items.len) |i| {
                const lo = ps.label_offsets.items[i];
                const hi = ps.label_offsets.items[i + 1];
                try self.multi_label_bytes.appendSlice(self.gpa, ps.label_bytes.items[lo..hi]);
                try self.multi_label_offsets.append(self.gpa, @intCast(self.multi_label_bytes.items.len));
            }
        }
    }
};

pub const Finalized = struct {
    row_infos: std.ArrayList(RowInfo),
    x0_pool: std.ArrayList(u8),
    x1_pool: std.ArrayList(u8),

    pub fn deinit(self: *Finalized, gpa: Allocator) void {
        self.row_infos.deinit(gpa);
        self.x0_pool.deinit(gpa);
        self.x1_pool.deinit(gpa);
    }
};

// Append a row's byte-stride sample stream into the shared pool and return its
// starting BYTE offset. The samples are tide's raw byte run (one sample =
// bytes_per_sample consecutive bytes), so this is a plain memcpy.
fn packRow(pool: *std.ArrayList(u8), gpa: Allocator, bytes: []const u8) !u32 {
    const start_byte: u32 = @intCast(pool.items.len);
    try pool.appendSlice(gpa, bytes);
    return start_byte;
}

pub fn finalize(scene: *Scene, gpa: Allocator) !Finalized {
    var max_row: usize = 0;
    for (scene.rows, 0..) |r, idx| {
        if (r.bit_width != 0) max_row = idx;
    }
    const row_count: usize = if (scene.rows[0].bit_width != 0 or max_row > 0) max_row + 1 else 0;

    var row_infos: std.ArrayList(RowInfo) = .empty;
    errdefer row_infos.deinit(gpa);
    var x0: std.ArrayList(u8) = .empty;
    errdefer x0.deinit(gpa);
    var x1: std.ArrayList(u8) = .empty;
    errdefer x1.deinit(gpa);

    try row_infos.ensureTotalCapacity(gpa, row_count);

    var i: usize = 0;
    while (i < row_count) : (i += 1) {
        const r = scene.rows[i];
        if (r.bit_width == 0) {
            row_infos.appendAssumeCapacity(.{ .x0_offset = 0, .x1_offset = 0, .bytes_per_sample = 0, .segment_start = 0, .flags = 0, .y_offset = 0, .height = 0 });
            continue;
        }
        const bps = bytesPerSample(r.bit_width);
        const off0 = try packRow(&x0, gpa, r.lsbs.items);
        const off1 = try packRow(&x1, gpa, r.msbs.items);
        row_infos.appendAssumeCapacity(.{ .x0_offset = off0, .x1_offset = off1, .bytes_per_sample = bps, .segment_start = r.segment_start, .flags = 0, .y_offset = 0, .height = 0 });
    }

    // WebGPU writeBuffer needs a 4-byte-multiple size, and the shader reads the
    // pools as array<u32>; pad each tail to a word boundary with zeros (inert in
    // the OR-fold — they sit past every sample's byte run). One pad per pool, not
    // per sample, so inter-row byte offsets are unaffected.
    while (x0.items.len % 4 != 0) try x0.append(gpa, 0);
    while (x1.items.len % 4 != 0) try x1.append(gpa, 0);

    // Shader invariant: every segment's row index must point to a populated
    // RowInfo (bytes_per_sample > 0). decodeSample's loop runs bytes_per_sample
    // iterations — 0 would leave the decoded value undefined. Caught here once
    // at scene-build, not per-frame on the GPU.
    for ([_][]const PackedSegment{ scene.multi.items, scene.single.items }) |segs| {
        for (segs) |s| {
            const row = s.row_flags & 0xffff;
            std.debug.assert(row < row_infos.items.len);
            std.debug.assert(row_infos.items[row].bytes_per_sample > 0);
        }
    }

    return .{ .row_infos = row_infos, .x0_pool = x0, .x1_pool = x1 };
}

pub fn packSegmentsInto(dest: []u8, segs: []const PackedSegment) void {
    std.debug.assert(dest.len == segs.len * PACKED_SEGMENT_BYTES);
    @memcpy(dest, std.mem.sliceAsBytes(segs));
}

pub fn packRowInfosInto(dest: []u8, infos: []const RowInfo) void {
    std.debug.assert(dest.len == infos.len * ROW_INFO_BYTES);
    @memcpy(dest, std.mem.sliceAsBytes(infos));
}

File: native/src/segments.rs
//! The GPU wire format, and the pooling that assembles a scene into it.
//!
//! Three layouts are a binary contract with WGSL and must not drift:
//! [`PackedSegment`] (12 bytes) mirrors `digital.wgsl`'s segment storage buffer,
//! [`RowInfo`] (28 bytes) mirrors the `RowInfo` struct in both `digital.wgsl` and
//! `labels.wgsl`, and the two pools are byte-strided at `ceil(bit_width / 8)`.
//!
//! The pools carry tide's `min` plane and its unknown mask (`min ^ max`), which
//! is all the shader reads: it OR-folds each sample to "is the value non-zero"
//! and "does it hold any x or z", and picks a line level or a crosshatch from
//! that. Telling x from z needs the third plane and happens CPU-side, in
//! [`label`] and in the renderer's value column.

use crate::label::{self, EnumEntry, Radix};
use tide_core::metadata::Width;

pub const FLAG_SHADE: u32 = 1 << 16;
pub const FLAG_RIGHT_EDGE: u32 = 1 << 17;
pub const FLAG_RISING_EDGE: u32 = 1 << 18;
pub const FLAG_FALLING_EDGE: u32 = 1 << 19;
pub const FLAG_MUTE: u32 = 1 << 20;
pub const FLAG_RISING_EDGE_LEFT: u32 = 1 << 21;
pub const FLAG_FALLING_EDGE_LEFT: u32 = 1 << 22;

/// Timing plus the row index and edge/shade bits. Values live in the shared
/// pools, addressed through [`RowInfo`]: a segment's sample index within its row
/// is `instance_index - RowInfo::segment_start`.
#[derive(Clone, Copy, Default, PartialEq, Eq, Debug)]
#[repr(C)]
pub struct PackedSegment {
    pub t_start: u32,
    pub t_end: u32,
    /// `[15:0]` row index, `[22:16]` the `FLAG_*` bits above.
    pub row_flags: u32,
}

/// The shader reads three words per segment; drifting the stride silently
/// misreads the whole buffer.
const _: () = assert!(size_of::<PackedSegment>() == 12);

/// Per-row metadata, indexed by row.
#[derive(Clone, Copy, Default, PartialEq, Eq, Debug)]
#[repr(C)]
pub struct RowInfo {
    /// Byte offset into the value pool.
    pub x0_offset: u32,
    /// Byte offset into the unknown-mask pool.
    pub x1_offset: u32,
    /// `ceil(bit_width / 8)`, the same stride as tide's `Type::bytes`. Zero marks
    /// a row nothing was placed at.
    pub bytes_per_sample: u32,
    /// This row's first instance index within its pipeline.
    pub segment_start: u32,
    /// Per-row render flags (bit 0 dims the row). Always emitted as zero: the
    /// renderer writes these straight into the GPU buffer on an eye toggle, with
    /// no repack.
    pub flags: u32,
    /// Row top in canvas space, CSS px as f32 bits. Always emitted as zero — the
    /// renderer owns live layout, same as `flags`.
    pub y_offset: u32,
    /// Row height, CSS px as f32 bits. Always emitted as zero.
    pub height: u32,
}

/// Seven words per row, mirrored in `digital.wgsl` and `labels.wgsl`.
const _: () = assert!(size_of::<RowInfo>() == 28);

/// Bounded by the 16-bit row field in `PackedSegment::row_flags`. The renderer's
/// `gpu/colors.ts` `MAX_ROWS` must match.
pub const MAX_ROWS: u32 = 65535;

/// Bytes one sample of a `width`-bit signal occupies in the pools.
pub fn bytes_per_sample(width: Width) -> u32 {
    debug_assert!(width >= 1);
    width.div_ceil(8)
}

/// `ceil(width / 32)`, the word count used only by the CPU value path
/// (`getValueAt`), which keeps its word-array shape independently of the pools.
pub fn words_per_sample(width: Width) -> u32 {
    debug_assert!(width >= 1);
    width.div_ceil(32)
}

/// One packed signal, independent of the row it lands on: the row bits of every
/// `row_flags` are zero until placement.
///
/// `values`/`unknowns` hold `segments.len() * bytes_per_sample` bytes.
/// `label_offsets`
/// holds `segments.len() + 1` prefix offsets once any label is pushed, so label
/// `i` is `label_bytes[offsets[i]..offsets[i + 1]]`; it stays empty for an
/// unlabeled (binary) row.
#[derive(Default)]
pub struct PackedSignal {
    pub is_multi: bool,
    pub bit_width: Width,
    pub segments: Vec<PackedSegment>,
    pub values: Vec<u8>,
    pub unknowns: Vec<u8>,
    pub label_bytes: Vec<u8>,
    pub label_offsets: Vec<u32>,
}

impl PackedSignal {
    pub fn new(is_multi: bool, bit_width: Width) -> PackedSignal {
        PackedSignal {
            is_multi,
            bit_width,
            ..PackedSignal::default()
        }
    }

    /// Appends one transition's timing and flags. `flags` must carry no row bits.
    pub fn push_segment(&mut self, t_start: u32, t_end: u32, flags: u32) {
        self.segments.push(PackedSegment {
            t_start,
            t_end,
            row_flags: flags,
        });
    }

    /// Copies a chunk's whole value plane into the pools, deriving the unknown
    /// mask as it goes. Call once, after every `push_segment`: the i-th
    /// `bytes_per_sample` run then lines up with segment i.
    pub fn set_samples(&mut self, min: &[u8], max: &[u8]) {
        debug_assert_eq!(min.len(), max.len(), "plane lengths");
        self.values.extend_from_slice(min);
        self.unknowns
            .extend(min.iter().zip(max).map(|(low, high)| low ^ high));
    }

    /// Appends this segment's value label. Call once per `push_segment`, in
    /// order. A muted segment writes no bytes but still takes an offset slot, so
    /// label i stays aligned with segment i.
    pub fn push_label(
        &mut self,
        sample: (&[u8], &[u8], &[u8]),
        radix: Radix,
        enums: &[EnumEntry],
        muted: bool,
    ) {
        if self.label_offsets.is_empty() {
            self.label_offsets.push(0);
        }
        if !muted {
            let (min, max, z) = sample;
            let width = self.bit_width;
            label::format_value(&mut self.label_bytes, min, max, z, width, radix, enums);
        }
        self.label_offsets.push(self.label_bytes.len() as u32);
    }
}

/// Accumulates one row's samples while a scene is being assembled.
#[derive(Default)]
struct RowAccum {
    /// Zero marks an unused row.
    bit_width: Width,
    segment_start: u32,
    started: bool,
    /// Samples pushed; the pools hold `count * bytes_per_sample` bytes each.
    count: u32,
    values: Vec<u8>,
    unknowns: Vec<u8>,
}

/// Every placed signal, split by pipeline, plus the per-row sample accumulators.
///
/// Labels are routed to the stream matching the signal's pipeline. The single
/// stream carries an entry for *every* single segment — empty for the unlabeled
/// binary rows — so the label batch can index labels by segment index.
#[derive(Default)]
pub struct Scene {
    pub multi: Vec<PackedSegment>,
    pub single: Vec<PackedSegment>,
    pub multi_label_bytes: Vec<u8>,
    pub multi_label_offsets: Vec<u32>,
    pub single_label_bytes: Vec<u8>,
    pub single_label_offsets: Vec<u32>,
    /// Grown on demand to the highest placed row plus one. Gaps stay default and
    /// finalize emits an empty `RowInfo` for them.
    rows: Vec<RowAccum>,
}

/// Which pipeline a signal's segments join.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Pipeline {
    Single,
    Multi,
}

impl Scene {
    pub fn new() -> Scene {
        Scene::default()
    }

    /// Places an already packed signal at `row`: its segments join the pipeline
    /// its radix chose, its samples join the row's pool, and its labels join the
    /// matching label stream.
    ///
    /// # Panics
    /// If `row` is out of range, or a signal was already placed there — one
    /// signal fills a row, contiguously.
    pub fn push_packed_signal(&mut self, row: u32, signal: &PackedSignal) {
        assert!(row < MAX_ROWS, "row {row} past MAX_ROWS");
        // An empty signal contributes no row data, so the row stays unpopulated
        // and finalize emits a zeroed RowInfo for it.
        if signal.segments.is_empty() {
            return;
        }

        let index = row as usize;
        if self.rows.len() <= index {
            self.rows.resize_with(index + 1, RowAccum::default);
        }

        let pipeline = if signal.is_multi {
            Pipeline::Multi
        } else {
            Pipeline::Single
        };
        let target = match pipeline {
            Pipeline::Multi => &mut self.multi,
            Pipeline::Single => &mut self.single,
        };

        let accum = &mut self.rows[index];
        assert!(!accum.started, "row {row} already holds a signal");
        accum.bit_width = signal.bit_width;
        accum.segment_start = target.len() as u32;
        accum.started = true;

        target.extend(signal.segments.iter().map(|segment| PackedSegment {
            t_start: segment.t_start,
            t_end: segment.t_end,
            row_flags: (segment.row_flags & !0xffff) | (row & 0xffff),
        }));

        // One signal fills a row, so the accumulator starts empty and this is a
        // single copy of the whole byte run.
        accum.values.extend_from_slice(&signal.values);
        accum.unknowns.extend_from_slice(&signal.unknowns);
        accum.count += signal.segments.len() as u32;
        debug_assert_eq!(
            accum.values.len() as u32,
            accum.count * bytes_per_sample(accum.bit_width),
            "row {row} sample stride drift",
        );

        let (bytes, offsets) = match pipeline {
            Pipeline::Multi => (&mut self.multi_label_bytes, &mut self.multi_label_offsets),
            Pipeline::Single => (&mut self.single_label_bytes, &mut self.single_label_offsets),
        };
        if offsets.is_empty() {
            offsets.push(0);
        }
        let labeled = !signal.label_offsets.is_empty();
        for i in 0..signal.segments.len() {
            if labeled {
                let lo = signal.label_offsets[i] as usize;
                let hi = signal.label_offsets[i + 1] as usize;
                bytes.extend_from_slice(&signal.label_bytes[lo..hi]);
            }
            offsets.push(bytes.len() as u32);
        }
    }

    /// Concatenates the per-row sample runs into the two shared pools and emits
    /// one `RowInfo` per row up to the highest populated one.
    ///
    /// # Panics
    /// If a segment names a row with no samples behind it. `decodeSample` in the
    /// shader loops `bytes_per_sample` times, so a zero there would leave the
    /// decoded value undefined; catching it once at scene build beats a garbled
    /// frame.
    pub fn finalize(&self) -> Finalized {
        let row_count = self
            .rows
            .iter()
            .rposition(|row| row.bit_width != 0)
            .map_or(0, |index| index + 1);

        let mut row_infos = Vec::with_capacity(row_count);
        let mut x0_pool = Vec::new();
        let mut x1_pool = Vec::new();

        for row in &self.rows[..row_count] {
            if row.bit_width == 0 {
                row_infos.push(RowInfo::default());
                continue;
            }
            let x0_offset = x0_pool.len() as u32;
            x0_pool.extend_from_slice(&row.values);
            let x1_offset = x1_pool.len() as u32;
            x1_pool.extend_from_slice(&row.unknowns);
            row_infos.push(RowInfo {
                x0_offset,
                x1_offset,
                bytes_per_sample: bytes_per_sample(row.bit_width),
                segment_start: row.segment_start,
                ..RowInfo::default()
            });
        }

        // WebGPU's writeBuffer needs a four-byte multiple and the shader binds the
        // pools as array<u32>. One pad per pool, not per row, so inter-row byte
        // offsets are untouched and the zeros sit past every sample's run.
        x0_pool.resize(x0_pool.len().next_multiple_of(4), 0);
        x1_pool.resize(x1_pool.len().next_multiple_of(4), 0);

        for segment in self.multi.iter().chain(&self.single) {
            let row = (segment.row_flags & 0xffff) as usize;
            assert!(
                row < row_infos.len(),
                "segment names row {row} with no info"
            );
            assert!(
                row_infos[row].bytes_per_sample > 0,
                "segment names unpopulated row {row}",
            );
        }

        Finalized {
            row_infos,
            x0_pool,
            x1_pool,
        }
    }
}

/// The scene's per-row metadata and shared sample pools, ready for the GPU.
pub struct Finalized {
    pub row_infos: Vec<RowInfo>,
    pub x0_pool: Vec<u8>,
    pub x1_pool: Vec<u8>,
}

/// A `#[repr(C)]` record array as the bytes the GPU reads. Both layouts are
/// plain `u32` fields, so this is the host's little-endian representation with no
/// padding to skip.
///
/// # Safety
/// `T` must be a `#[repr(C)]` struct of `u32` fields only.
unsafe fn record_bytes<T>(records: &[T]) -> &[u8] {
    // SAFETY: the caller guarantees T is a padding-free struct of u32s, so every
    // byte of the array is initialized and any bit pattern is valid as u8.
    unsafe { std::slice::from_raw_parts(records.as_ptr().cast::<u8>(), size_of_val(records)) }
}

pub fn segment_bytes(segments: &[PackedSegment]) -> &[u8] {
    // SAFETY: PackedSegment is #[repr(C)] with three u32 fields.
    unsafe { record_bytes(segments) }
}

pub fn row_info_bytes(rows: &[RowInfo]) -> &[u8] {
    // SAFETY: RowInfo is #[repr(C)] with seven u32 fields.
    unsafe { record_bytes(rows) }
}

/// A `u32` slice as the bytes the GPU reads, for the label offset arrays.
pub fn word_bytes(words: &[u32]) -> &[u8] {
    // SAFETY: u32 has no padding and every bit pattern is a valid u8.
    unsafe { std::slice::from_raw_parts(words.as_ptr().cast::<u8>(), size_of_val(words)) }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// A signal of `count` one-byte-wide segments, all flagless.
    fn signal(is_multi: bool, width: Width, count: usize) -> PackedSignal {
        let mut signal = PackedSignal::new(is_multi, width);
        let stride = bytes_per_sample(width) as usize;
        for i in 0..count {
            signal.push_segment(i as u32 * 10, (i as u32 + 1) * 10, 0);
        }
        let values = vec![0xaa; count * stride];
        signal.set_samples(&values, &values);
        signal
    }

    #[test]
    fn record_layouts_match_the_shader() {
        // digital.wgsl reads three words per segment and seven per row; drifting
        // either stride silently misreads every buffer. The struct sizes are
        // pinned at compile time, so this covers the byte views over them.
        assert_eq!(12, segment_bytes(&[PackedSegment::default()]).len());
        assert_eq!(28, row_info_bytes(&[RowInfo::default()]).len());
    }

    #[test]
    fn placement_ors_the_row_into_every_segment() {
        let mut scene = Scene::new();
        scene.push_packed_signal(3, &signal(false, 1, 2));
        assert_eq!(2, scene.single.len());
        for segment in &scene.single {
            assert_eq!(3, segment.row_flags & 0xffff);
        }

        // Flags above bit 15 survive; the row replaces only the low half.
        let mut flagged = signal(false, 1, 1);
        flagged.segments[0].row_flags = FLAG_SHADE | FLAG_RIGHT_EDGE;
        scene.push_packed_signal(4, &flagged);
        let placed = scene.single[2];
        assert_eq!(4, placed.row_flags & 0xffff);
        assert_eq!(FLAG_SHADE | FLAG_RIGHT_EDGE, placed.row_flags & !0xffff);
    }

    #[test]
    fn rows_are_pooled_in_order_with_gaps_left_empty() {
        let mut scene = Scene::new();
        // Row 0 is left unplaced, so it must still get a zeroed RowInfo.
        scene.push_packed_signal(1, &signal(true, 16, 2));
        scene.push_packed_signal(2, &signal(true, 8, 3));

        let final_scene = scene.finalize();
        assert_eq!(3, final_scene.row_infos.len());
        assert_eq!(RowInfo::default(), final_scene.row_infos[0]);

        let (first, second) = (final_scene.row_infos[1], final_scene.row_infos[2]);
        assert_eq!(2, first.bytes_per_sample);
        assert_eq!(0, first.x0_offset);
        assert_eq!(0, first.segment_start);
        assert_eq!(1, second.bytes_per_sample);
        // Row 1 wrote two 2-byte samples, so row 2 starts four bytes in.
        assert_eq!(4, second.x0_offset);
        assert_eq!(2, second.segment_start);
    }

    #[test]
    fn pools_are_padded_to_a_word() {
        let mut scene = Scene::new();
        // Three one-byte samples is not a multiple of four.
        scene.push_packed_signal(0, &signal(false, 8, 3));
        let final_scene = scene.finalize();
        assert_eq!(4, final_scene.x0_pool.len());
        assert_eq!(4, final_scene.x1_pool.len());
        assert_eq!(0, final_scene.x0_pool[3]);
    }

    #[test]
    fn every_segment_gets_a_label_offset_even_when_unlabeled() {
        let mut scene = Scene::new();
        // A binary row pushes no labels at all, but the batch still indexes
        // labels by segment index, so the offsets must keep pace.
        scene.push_packed_signal(0, &signal(false, 1, 3));
        assert_eq!(vec![0, 0, 0, 0], scene.single_label_offsets);
        assert!(scene.single_label_bytes.is_empty());

        let mut labeled = PackedSignal::new(true, 8);
        for i in 0..2 {
            labeled.push_segment(i, i + 1, 0);
            labeled.push_label((&[0xab], &[0xab], &[0x00]), Radix::Hex, &[], false);
        }
        labeled.set_samples(&[0xab, 0xab], &[0xab, 0xab]);
        scene.push_packed_signal(1, &labeled);
        assert_eq!(vec![0, 4, 8], scene.multi_label_offsets);
        assert_eq!(b"0xAB0xAB", scene.multi_label_bytes.as_slice());
    }

    #[test]
    fn an_empty_signal_leaves_its_row_unpopulated() {
        let mut scene = Scene::new();
        scene.push_packed_signal(0, &PackedSignal::new(false, 8));
        // Nothing placed anywhere, so there are no rows to describe.
        assert_eq!(0, scene.finalize().row_infos.len());
    }

    #[test]
    #[should_panic(expected = "already holds a signal")]
    fn two_signals_cannot_share_a_row() {
        let mut scene = Scene::new();
        scene.push_packed_signal(0, &signal(false, 1, 1));
        scene.push_packed_signal(0, &signal(false, 1, 1));
    }
}
